hashring/hashring.go

```diff
@@ -89,32 +99,47 @@ func New(hashfunc func([]byte) uint32, replicaPoints int) *HashRing {
 
 // Checksum returns the checksum of all stored servers in the HashRing
 // Use this value to find out if the HashRing is mutated.
-func (r *HashRing) Checksum() uint32 {
+func (r *HashRing) Checksum() (checksum uint32) {
 	r.RLock()
-	checksum := r.checksum
+	checksum = r.legacyChecksum
 	r.RUnlock()
-	return checksum
+	return
 }
 
-// computeChecksum computes checksum of all servers in the ring.
-// This function isn't thread-safe, only call it when the HashRing is locked.
-func (r *HashRing) computeChecksumNoLock() {
-	addresses := r.copyServersNoLock()
-	sort.Strings(addresses)
-	bytes := []byte(strings.Join(addresses, ";"))
-	old := r.checksum
-	r.checksum = farm.Fingerprint32(bytes)
-
-	if r.checksum != old {
+func (r *HashRing) Checksums() (checksums map[string]uint32) {
+	r.RLock()
+	// even though the map is immutable the pointer to it is not so it requires
```

Comment:
Should we add a check if checksums are already calculated and compute them if not?